e3nn/experimental/_full_tp.py

```diff
@@ -0,0 +1,113 @@
+from e3nn.util.datatypes import Path, Chunk
+from typing import Union, Callable, Optional
+from e3nn import o3
+
+import torch
+from torch import nn
+import numpy as np
+
+
+def _prepare_inputs(input1, input2):
+    dtype = torch.promote_types(input1.dtype, input2.dtype)
+    input1 = input1.to(dtype=dtype)
+    input2 = input2.to(dtype=dtype)
+    leading_shape = torch.broadcast_shapes(input1.shape[:-1], input2.shape[:-1])
+    input1 = input1.broadcast_to(leading_shape + (-1,))
+    input2 = input2.broadcast_to(leading_shape + (-1,))
+    return input1, input2, leading_shape
+
+
+class IrrepsContext:
+    def __init__(self, irreps_class):
+        self._Irreps = irreps_class
+
+    def clebsch_gordan(self, ir_1, ir_2, ir_out):
+        raise NotImplementedError
+
+    def path_hash(self, ir_1, ir_2, ir_out):
+        raise NotImplementedError
+
+
+class O3Context(IrrepsContext):
+    def __init__(self):
+        super().__init__(o3.Irreps)
+
+    def clebsch_gordan(self, ir_1, ir_2, ir_out):
+        return o3.wigner_3j(ir_1.l, ir_2.l, ir_out.l)
+
+    def path_hash(self, ir_1, ir_2, ir_out):
+        return (ir_1.l, ir_2.l, ir_out.l)
```

Comment:
What is the goal of this class?

Reply:
Ultimately want it to look like [this](https://github.com/blondegeek/e3nn/pull/1/files#diff-97b92416caa881d4af264edcc4b5ced53aac858c902d6924aecf94327a2c93e1R20-R50).